Multi-domain: Add error message if domain removal or add fails

DustyReagan opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/4704

## Proposed Changes

* Adds a dismissable error message to the top of the domain selection if the domain fails to be removed or added to the mini-cart due to an otherwise uncaught error.

<img width="1718" alt="Screenshot 2023-12-06 at 11 01 58 AM" src="https://github.com/Automattic/wp-calypso/assets/140841/7f5d2ebe-fcbd-4d22-b138-26799d165e31">

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Sandbox the API and edit the API to return an error on adding or removing a domain fbhepr%2Skers%2Sjcpbz%2Schoyvp.ncv%2Serfg%2Sjcpbz%2Qwfba%2Qraqcbvagf%2Spynff.jcpbz%2Qwfba%2Qncv%2Qzr%2Qfubccvat%2Qpneg%2Qraqcbvagf.cuc%3Se%3Q8qsooq5q%23155-og
* Try adding or removing a domain from /start/domains
* Make sure /start/domains still works when there are no errors
* Make sure it works on /domains
* Make sure it looks ok on mobile



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

DustyReagan committed:
Add error message if domain removal or add fails

candy02058912 committed:
Allow domains to be selected after failed message

candy02058912 committed:
Revert "Allow domains to be selected after failed message"

This reverts commit a6a8457478a45cc810b8b039af1d0e19d07753a8.

candy02058912 reviewed (commented):
Took a stab at making the Select button available after the error, but ultimately found out that it caused more bugs, hence the revert.

candy02058912 commented:
I'm having mixed feelings about what we have currently.
A workaround might include adding something like "please close the error message to proceed" or it feels "frozen" with all the buttons disabled. What do you think?

![disabled](https://github.com/Automattic/wp-calypso/assets/6586048/14b7e766-9d57-4893-b766-089081c26cca)

The notice itself looks good on desktop/mobile.

DustyReagan committed:
Move error location and reload shoppingcart from server on error

DustyReagan commented:
> I'm having mixed feelings about what we have currently. A workaround might include adding something like "please close the error message to proceed" or it feels "frozen" with all the buttons disabled. What do you think?

Agreed! I think I figured out how to unfreeze the buttons with `shoppingCartManager.reloadFromServer()`. I also updated the error copy and moved the error to the location @paulopmt1 suggested here: p1701800717415359/1701729597.564039-slack-CKZHG0QCR

candy02058912 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -655,7 +656,19 @@ export class RenderDomainsStep extends Component {
+							'Sorry, there was an error adding that domain. Please refresh the page to try again.'
+						),
+					} );
+					this.props.shoppingCartManager.reloadFromServer();
```
Great job on discovering this TIL!

candy02058912 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -723,9 +736,16 @@ export class RenderDomainsStep extends Component {
 				.catch( () => {
+					this.setState( {
+						replaceDomainFailedMessage: this.props.translate(
+							'Sorry, we had a problem adding that domain. Please refresh the page and try again.'
```
```suggestion
							'Sorry, we had a problem removing that domain. Please refresh the page and try again.'
```

candy02058912 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -655,7 +656,19 @@ export class RenderDomainsStep extends Component {
+				.catch( () => {
+					this.setState( {
+						replaceDomainFailedMessage: this.props.translate(
+							'Sorry, there was an error adding that domain. Please refresh the page to try again.'
```
As the buttons are unfrozen now, perhaps:
```suggestion
							'Sorry, there was an error adding that domain. Please try again later.'
```

We're not sure if refreshing the page would help with their issue.
What do you think?

candy02058912 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -1330,9 +1357,11 @@ export class RenderDomainsStep extends Component {
-					<Notice status="is-error" showDismiss={ false }>
-						{ this.props.step.errors.message }
-					</Notice>
+					{ this.props.step && 'invalid' === this.props.step.status && (
```
Curious about what this line does, because on line 1357 it checked `this.props.step && 'invalid' === this.props.step.status` so doesn't this always resolve to `true`?

DustyReagan replied:
Good catch, thanks! That's dead code. 😅

candy02058912 reviewed (commented):
Some minor comments around the copy, and a question about the conditions.
Overall I like this version better than yesterdays!

DustyReagan committed:
Apply copy suggestions

Co-authored-by: Candy Tsai <candy02058912@gmail.com>

DustyReagan committed:
Copy update and remove dead code

paulopmt1 reviewed (approved):
Nice work on this @DustyReagan!

DustyReagan committed:
Only show uncaught errors and reload the temp cart

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9993259

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @DustyReagan for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9993259).
